File: src/leetcode/recursion/Binary_Tree_Maximum_Path_Sum_124.java
package leetcode.recursion;

/**
 * Created by sunbo_000 on 10/15/2016.
 */

/*
    https://leetcode.com/problems/binary-tree-maximum-path-sum/
 */
public class Binary_Tree_Maximum_Path_Sum_124 {

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public int maxPathSum(TreeNode root) {
        int[] max = {Integer.MIN_VALUE};
        rec(root, max);
        return max[0];
    }

    private int rec(TreeNode node, int[] max) {

        if (node == null) return 0;


        int left = rec(node.left, max);
        int right = rec(node.right, max);

        int arch = left + node.val + right;

        int maxToParent = Math.max(node.val, Math.max(left, right) + node.val);
        int maxPath = Math.max(arch, maxToParent);
        max[0] = Math.max(max[0], maxPath);
        return maxToParent;
    }

    public static void main(String[] args) {
        TreeNode one = new TreeNode(1);
        TreeNode two = new TreeNode(1);
        TreeNode three = new TreeNode(1);
        Binary_Tree_Maximum_Path_Sum_124 solution = new Binary_Tree_Maximum_Path_Sum_124();
        System.out.println(solution.maxPathSum(one));
        one.left = two;
        one.right = three;
        System.out.println(solution.maxPathSum(one));
    }
}

File: src/leetcode_javascript/Binary_Tree_Maximum_Path_Sum_124.js
/**
 * 124 Binary Tree Maximum Path Sum
 *
 * Given a binary tree, find the maximum path sum.

  For this problem, a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections. The path does not need to go through the root.

  For example:
  Given the below binary tree,

         1
        / \
       2   3
  Return 6.
 *
 * https://leetcode.com/problems/binary-tree-maximum-path-sum/
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    var max = Number.MIN_SAFE_INTEGER;
    if (!root) {
        return;
    }

    function helper(node) {
        if (node === null) {
            return 0;
        }
        var left = Math.max(0, helper(node.left));
        var right = Math.max(0, helper(node.right));

        max = Math.max(max, left + right + node.val);

        return Math.max(left, right) + node.val;
    }
    helper(root);
    return max;
};